Lib/compression/zstd/_zstdfile.py

```diff
@@ -0,0 +1,378 @@
+                raise TypeError("level argument should only be passed when writing.")
+            mode_code = _MODE_READ
+        elif mode in ("w", "wb", "a", "ab", "x", "xb"):
+            if not isinstance(level, (type(None), int)):
+                raise TypeError("level argument should be an int object.")
+            if not isinstance(options, (type(None), dict)):
+                raise TypeError("options argument should be an dict object.")
+            mode_code = _MODE_WRITE
+            self._compressor = ZstdCompressor(
+                level=level, options=options, zstd_dict=zstd_dict
+            )
+            self._pos = 0
+        else:
+            raise ValueError(f"Invalid mode: {mode!r}")
+
+        # File object
+        if isinstance(filename, (str, bytes, PathLike)):
+            if "b" not in mode:
+                mode += "b"
+            self._fp = builtins.open(filename, mode)
+            self._closefp = True
+        elif hasattr(filename, "read") or hasattr(filename, "write"):
+            self._fp = filename
+        else:
+            raise TypeError("filename must be a str, bytes, file or PathLike object")
+        self._mode = mode_code
+
+        if self._mode == _MODE_READ:
+            raw = self._READER_CLASS(
+                self._fp,
+                ZstdDecompressor,
+                trailing_error=ZstdError,
+                zstd_dict=zstd_dict,
+                options=options,
+            )
+            self._buffer = io.BufferedReader(raw)
+
+    def close(self):
+        """Flush and close the file.
+
+        May be called more than once without error. Once the file is
+        closed, any other operation on it will raise a ValueError.
+        """
+        # Nop if already closed
+        if self._fp is None:
+            return
+        try:
+            if self._mode == _MODE_READ:
+                if hasattr(self, "_buffer") and self._buffer:
+                    self._buffer.close()
+                    self._buffer = None
+            elif self._mode == _MODE_WRITE:
+                self.flush(self.FLUSH_FRAME)
+                self._compressor = None
+        finally:
+            self._mode = _MODE_CLOSED
+            try:
+                if self._closefp:
+                    self._fp.close()
+            finally:
+                self._fp = None
+                self._closefp = False
+
+    def write(self, data):
+        """Write a bytes-like object to the file.
+
+        Returns the number of uncompressed bytes written, which is
+        always the length of data in bytes. Note that due to buffering,
+        the file on disk may not reflect the data written until .flush()
+        or .close() is called.
+        """
+        self._check_can_write()
+        if isinstance(data, (bytes, bytearray)):
+            length = len(data)
+        else:
+            # accept any data that supports the buffer protocol
+            data = memoryview(data)
+            length = data.nbytes
```

Comment:
can this use the nbytes helper?

Reply:
Unfortunately that would lead to circular import. I could split `_nbytes` out into it's own file I guess?

Reply:
Would moving the helper into `_zstdfile` solve the circular import?

Reply:
Moving to this file sounds good